### PR TITLE
vtk: update package using changed variable names

### PR DESCRIPTION
**Pull request by glennpj**
This PR updates the vtk package to use new variable names and adds some
external dependencies.

- add version 1.4.2 to gl2ps package. This is needed to use gl2ps as a
  dependency.
- new package: utf8. This can now be sued as a dependency for version 9.
- use the package provided FindHDF5.cmake for versions up to 8. Version
  9 already uses an internal version.
- adjust the hdf dependencies to match the ~mpi and +mpi variants
- add external dependencies when possible
    - pugixml
    - libogg
    - libtheora
    - utf8
    - gl2ps
    - proj
- turn off configuring MPI if ~mpi
- use new variables for version 9
- remove unused variables depending on version

### CODE REVIEW

File: var/spack/repos/builtin/packages/seacas/package.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ class Seacas(CMakePackage):
     # Everything should be compiled position independent (-fpic)
 
     depends_on('netcdf-c@4.8.0:+mpi+parallel-netcdf', when='+mpi')
-    depends_on('netcdf-c@4.8.0:~mpi', when='~mpi')
+    depends_on('netcdf-c@4.8.0:~mpi~parallel-netcdf', when='~mpi')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/seacas/package.py]: Question: is it not possible to use `netcdf-c+parallel-netcdf` with `~mpi`? I'm curious if we can still use the dependency built with `+parallel-netcdf` even if we don't want to use `+mpi` for `seacas`.

File: var/spack/repos/builtin/packages/utf8/package.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Utf8(Package):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/utf8/package.py]: Could this be updated to use the full package name: `Utf8Cpp` which should be installable like `spack install utf8-cpp`

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 3]
```diff
@@ -101,7 +105,8 @@ class Vtk(CMakePackage):
     depends_on('freetype')
     depends_on('glew')
     # set hl variant explicitly, similar to issue #7145
-    depends_on('hdf5+hl')
+    depends_on('hdf5+hl~mpi', when='~mpi')
+    depends_on('hdf5+hl+mpi', when='+mpi')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/vtk/package.py]: Does this definitely require `hdf5+mpi` when `+mpi`? I've encountered packages that build without `mpi` but can still use `hdf5+mpi`.

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 4]
```diff
@@ -218,14 +247,14 @@ def cmake_args(self):
             qt_bin = spec['qt'].prefix.bin
             qmake_exe = os.path.join(qt_bin, 'qmake')
 
-            cmake_args.extend([
-                # Enable Qt support here.
-                '-DVTK_QT_VERSION:STRING={0}'.format(qt_ver),
-                '-DQT_QMAKE_EXECUTABLE:PATH={0}'.format(qmake_exe),
-                '-DVTK_Group_Qt:BOOL=ON',
-            ])
             # https://github.com/martijnkoopman/Qt-VTK-viewer/blob/master/doc/Build-VTK.md
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/vtk/package.py:249]: It looks like this comment is supposed to be attached to the `else` now.

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 5]
```diff
@@ -147,38 +164,45 @@ def cmake_args(self):
...
-                '-DVTK_USE_SYSTEM_PUGIXML:BOOL=OFF',
+                '-DVTK_USE_SYSTEM_LIBRARIES:BOOL=ON',
+                '-DVTK_USE_SYSTEM_LIBHARU=OFF',
             ])
+            if spec.satisfies('@:8.0'):
+                cmake_args.append('-DVTK_USE_SYSTEM_GL2PS=OFF')
         else:
             cmake_args.extend([
-                '-DVTK_USE_SYSTEM_OGGTHEORA:BOOL=OFF',
+                '-DVTK_USE_EXTERNAL:BOOL=ON',
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/vtk/package.py:186]: (question: no change needed) Did this replace `VTK_USE_SYSTEM_LIBRARIES`?

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 6]
```diff
@@ -235,37 +264,54 @@ def cmake_args(self):
...
+                    "-DBOOST_NO_SYSTEM_PATHS:BOOL=ON",
+                    # This is needed because VTK has multiple FindBoost
+                    # and they stick to system boost if there's a system boost
+                    # installed with CMake
+                    "-DBoost_NO_BOOST_CMAKE:BOOL=ON",
+                    # The xdmf project does not export any CMake file...
+                    "-DVTK_USE_SYSTEM_XDMF3:BOOL=OFF",
+                    "-DVTK_USE_SYSTEM_XDMF2:BOOL=OFF"
+                ])
+            else:
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/vtk/package.py:302]: I see `HDF5_ROOT` is handled elsewhere here but not `BOOST_ROOT` - is that no-longer necessary for `@9:`?

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few requests and a few questions.

File: var/spack/repos/builtin/packages/seacas/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/seacas/package.py]: That is a good question and I think the answer is, "it is possible", because `parallel-netcdf`  does not depend on `netcdf-c`. However, the `netcdf-c` package recipe has the following:
```
 depends_on('mpi', when='+parallel-netcdf')
```
which makes the above dependency necessary in `seacas`. I was not inclined to mess with the `netcdf-c` packages in this PR.

File: var/spack/repos/builtin/packages/utf8/package.py (hunk 2 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/utf8/package.py]: Yes, it could be, I even had it that way at one point while working on this. I settled on `utf8` as that is what the zip file is called. The `spack create` command determined the name was `utf` so this one is odd.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 3 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py]: This is similar to the previous question about `mpi`. I set it as above to be consistent with how the `netcdf-c` package handles the `mpi` variant. However, I am pretty sure that I was still using the original concretizer when I developed this PR so the extra consistency helped reduce conflicts. With the clingo concretizer this is probably no longer relevant. So, reverting this change should be okay.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 4 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py:249]: No, the document that the link refers to has been updated over time for newer versions, at one point it referred to VTK-8. Perhaps I should add a comment to note that.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 5 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py:186]: Yes.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 6 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py:302]: Correct, `@9:` is able to find `boost`.

File: var/spack/repos/builtin/packages/utf8/package.py (hunk 2 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/utf8/package.py]: @scheibelp Can you clarify whether this should be `utf8-cpp` or `utf8cpp`? I ask because the naming for this package is all over the map. The project homepage refers to `utf8-cpp`. Other than that, the code for the software is referred to as:

- utfcpp
- utf8cpp
- utf8

The VTK project seems to prefer `utf8cpp` so that is what I would lean towards.

File: var/spack/repos/builtin/packages/seacas/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/seacas/package.py]: I reverted this change as the clingo concretizer handles this natively.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 3 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py]: I reverted this as well as the `netcdf` change in `seacas`. I tested with the `clingo` concretizer and as suspected, it resolves everything to `~mpi`.

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 4 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/vtk/package.py:249]: I added a comment for this in the package.

File: var/spack/repos/builtin/packages/utf8/package.py (hunk 2 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/utf8/package.py]: I prefer `utf8-cpp` (to achieve that, you would choose `Utf8Cpp` as the name of the Package class in the `package.py` file).

File: var/spack/repos/builtin/packages/seacas/package.py [hunk 7]
```diff
@@ -126,6 +126,8 @@ def cmake_args(self):
                 '-DTPL_ENABLE_MPI:BOOL=ON',
                 '-DMPI_BASE_DIR:PATH=%s'      % spec['mpi'].prefix,
             ])
+        else:
+            '-DTPL_ENABLE_MPI:BOOL=OFF',
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/seacas/package.py]: This trailing comma looks odd to me but I wouldn't be shocked if it is somehow valid Python - have you successfully installed with `~mpi` since making this change?

Comment (glennpj) [reply var/spack/repos/builtin/packages/seacas/package.py]: Yes, I did build it successfully. However, the stray comma is likely a
"copy --> paste --> (in)complete edit" error and does not belong there. I just added a commit to fix it.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
I was reading through this and noticed:

> use the package provided FindHDF5.cmake for versions up to 8. Version 9 already uses an internal version.

It looks like this PR sets `HDF5_ROOT` for `@9:`, i.e. that it uses a Spack-built HDF5 - that's a good thing IMO (generally we don't want a package to install its own dependencies) but I wanted to double check with you because that contradicts this line from the commit message.

**Comment by glennpj** [comment]
> I was reading through this and noticed:
> 
> > use the package provided FindHDF5.cmake for versions up to 8. Version 9 already uses an internal version.
> 
> It looks like this PR sets `HDF5_ROOT` for `@9:`, i.e. that it uses a Spack-built HDF5 - that's a good thing IMO (generally we don't want a package to install its own dependencies) but I wanted to double check with you because that contradicts this line from the commit message.

There is no contradiction, just unclear wording on my part due to lacking context. The comment is in reference to the cmake files/code used to find HDF5, not an internal implementation of HDF5. While there is a cmake file that is part of the cmake distribution for finding HDF5, it does not work well with the HL component. This is a known issue and VTK includes its own custom cmake code for finding `hdf5+hl` to address this. For `vtk@:8` however, it will try to use cmake's distributed code to find `hdf5+hl` for newer cmake versions, but it will fail to find `hdf5+hl`. I included a patch so `vtk@:8` will always use its own, older, cmake code rather than that distributed with cmake. Now, `vtk@9:` always uses its own cmake code so a patch is not needed. The respective `vtk@:8` and `vtk@9:` cmake codes use different methods to find `hdf5+hl`, and the `vtk@9:` code needs `HDF5_ROOT` to be set, while the `vtk@:8` code does not.

In summary:
- `vtk@:8` > needs a patch to use vtk provided cmake code but does not need `HDF5_ROOT` set
- `vtk@9:` > does not need a patch to use vtk provided cmake code but needs `HDF5_ROOT` set

**Comment by scheibelp** [comment]
@spackbot re-run pipeline
